### DISCUSSION TITLE
`MPIPoolExecutor` tries to the Python executable from another host

### DISCUSSION THREAD

**Discussion by winstxnhdw** [mpi4py/mpi4py#581] [Q&A] [answered]
I am posting this as a discussion because I am not confident that this is actually an issue with `mpi4py`.

Using the `MPIPoolExecutor` example from the documentation with the following command,

```bash
mpiexec -hosts 2 localhost another.host.local python test.py
```

```python
import math
import sys
from mpi4py.futures import MPIPoolExecutor, wait
from mpi4py.futures import get_comm_workers
from socket import gethostname

def compute_pi(n):
    # Access intracommunicator and synchronize
    comm = get_comm_workers()
    comm.Barrier()

    rank = comm.Get_rank()
    size = comm.Get_size()
    hostname = gethostname()
    print(f"I am rank {rank} from {hostname}!")

    # Local computation
    h = 1.0 / n
    s = 0.0
    for i in range(rank + 1, n + 1, size):
        x = h * (i - 0.5)
        s += 4.0 / (1.0 + x**2)
    pi_partial = s * h

    # Parallel reduce-to-all
    pi = comm.allreduce(pi_partial)

    # All workers return the same value
    return pi


if __name__ == '__main__':
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 256

    with MPIPoolExecutor(max_workers=10) as executor:
        # Submit exactly one callable per worker
        P = executor.num_workers
        fs = [executor.submit(compute_pi, n) for _ in range(P)]

        # Wait for all workers to finish
        wait(fs)

        # Get result from the first future object.
        # In this particular example, due to using reduce-to-all,
        # all the other future objects hold the same result value.
        pi = fs[0].result()
        print(
            f"pi: {pi:.16f}, error: {abs(pi - math.pi):.3e}",
            f"({n:d} intervals, {P:d} workers)",
        )
```

returns the following error:

```
ERROR: Error reported: failed to launch 'C:\Users\another.host.local\AppData\Local\Programs\Python\Python311\python.exe  "-R" "-m" "mpi4py.futures.server"' on MK-LPT-12223
```

`MK-LPT-12223` here is the device name of my localhost from which I am dispatching the MPI request. For some reason, MPI is trying to run the Python executable from `another.host.local` in my `localhost` device.

In contrast, if I were to just use the MPI primitives instead, I don't get this issue.

```python
import numpy as np
from mpi4py import MPI
from socket import gethostname
from sys import executable

# Initialize the MPI communicator
comm = MPI.COMM_WORLD

# Get the rank of the current process
rank = comm.Get_rank()

# Get the total number of processes
size = comm.Get_size()
hostname = gethostname()
print(f"{hostname=}")
print(f"{executable=}")
# Define the size of the array (let's take a large array for parallel computation)
array_size = 1000000

# Create a large array only on the root process (rank 0)
if rank == 0:
    data = np.arange(array_size, dtype="i")
else:
    data = None

# Determine the chunk size for each process
chunk_size = array_size // size

# Create a buffer to hold the subset of data for each process
local_data = np.empty(chunk_size, dtype="i")

# Scatter the data to all processes
comm.Scatter(data, local_data, root=0)

# Each process computes the sum of its chunk
local_sum = np.sum(local_data)

# Print local sum for each process (optional)
print(f"Process {rank}: Local sum = {local_sum}")

# Use MPI Reduce to sum up the results from all processes at the root process
total_sum = comm.reduce(local_sum, op=MPI.SUM, root=0)

# The root process (rank 0) prints the total sum
if rank == 0:
    print(f"Total sum = {total_sum}")

# Ensure synchronization
comm.Barrier()
```

Looking through the code base of `mpi4py`, I don't see any specific code that is expanding the Python binary path, so I don't think this is exactly an issue with `MPIPoolExecutor` but clearly there's some abstraction magic going on here.

I am using `mpi4py` 4.0.1.

**Comment by dalcinl** [comment, marked as answer]
> Looking through the code base of `mpi4py`, I don't see any specific code that is expanding the Python binary path, so I don't think this is a exactly an issue with `MPIPoolExecutor` but clearly there's some abstraction magic going on here.

mpi4py relies on [`sys.executable`](https://docs.python.org/3/library/sys.html#sys.executable) as the default Python binary. 
The default is handled here https://github.com/mpi4py/mpi4py/blob/master/src/mpi4py/futures/_core.py#L1046.
You can override the default, though , using the `python_exe` keyword argument as documented [here](https://mpi4py.readthedocs.io/en/stable/mpi4py.futures.html#mpi4py.futures.MPIPoolExecutor).

Maybe the root of the problem is that your filesystem in each host does not have the same directory structure. If that is the case, I don't have any tips other than not doing things that way. Otherwise, you should ask the MPI implementation developers how things should be handled in that case. I'm not a windows expert, but my wild guess is that passing `python_exe="python.exe"` as keyword arg to `MPIPoolExecutor` may be enough, assuming that on each host `python.exe` can be bound via PATH.

**Reply by winstxnhdw** [reply]
I am using the same directory structure, but where Python is installed on each machine is different. You are right. I’ll give that a try.

**Reply by winstxnhdw** [reply]
> mpi4py relies on [sys.executable](https://docs.python.org/3/library/sys.html#sys.executable) as the default Python binary.
The default is handled here https://github.com/mpi4py/mpi4py/blob/master/src/mpi4py/futures/_core.py#L1046.

Still, this doesn’t really explain why this mixup is happening. `sys.executable` should correctly resolve my Python binary and not use another hosts’. Now it really does seem like a `mpi4py` bug.

**Reply by dalcinl** [reply]
> I am using the same directory structure, but where Python is installed on each machine is different. 

That accounts as having a different directory structures. 

> Still, this doesn’t really explain why this mixup is happening. 

You have host A asking MPI to execute Python on host B. But host A makes such request using `sys.executable` as the full path to the Python binary. Then MPI goes to host B, using the full path he got from host A, and cannot find `python.exe`.

Long story short: mpi4py's default of using `sys.executable` is a very good guess for the vast majority of cases where a shared filesystem is used in compute nodes. Obviously, such default fails miserably if each compute node have Python installed in different locations. That seems to be your case, then it is clear to me things will not work. 
Now to the important point: how could you fix this problem? I believe using the `python_exe="python.exe"` argument may work, assuming your MPI can finds in each compute node the actual location of `python.exe` via the `PATH` environment variable. This of course requires that in each compute node `PATH` is properly set, such that you can type `python.exe` in a console and execute it without having to specify the full location.

**Reply by winstxnhdw** [reply]
> You have host A asking MPI to execute Python on host B. But host A makes such request using sys.executable as the full path to the Python binary. Then MPI goes to host B, using the full path he got from host A, and cannot find python.exe.

Except this isn’t what’s happening. In this case, host A should be my laptop (rank 0) and host B would be my other machine. If it’s what you said, we should see that host B is trying to use host A’s Python executable path. However, according to my error message, we see it the other way around; host A is trying to use host B’s Python executable path.

**Reply by dalcinl** [reply]
Maybe MPI is not mapping processes to compute nodes the way you think.
Are you using Intel MPI? Why are you passing "2" after the `-hosts` option to `mpiexec`? 

```console
mpiexec -hosts localhost,another.host.local python test.py
```

Again, I'm not used to work in Windows and never tried to do what you are trying to do on Windows.

PS: Did you try passing `python_exe="python.exe"` as keyword arg to `MPIPoolExecutor` ?

**Reply by winstxnhdw** [reply]
> Why are you passing "2" after the -hosts option to mpiexec?

I am using Microsoft MPI, and the `hosts` argument is structured like this.

```bash
hosts [number_of_hosts] [...hostname] # where the first host is rank 0
```

> PS: Did you try passing python_exe="python.exe" as keyword arg to MPIPoolExecutor ?

Yes, I just tried it. It works! Thank you for the help!

**Reply by dalcinl** [reply]
Microsoft MPI have become abandonware. You should really try to switch to Intel MPI.

**Reply by winstxnhdw** [reply]
> Yes, I just tried it. It works! Thank you for the help!

It took me a while to realise that although the Python binary now resolves properly and it seemed to work correctly, the executor was running entirely on my laptop and not distributing the work the my other machine as well. It seems that `MPIPoolExecutor` isn't automatically detecting the universe size from my hosts that I defined.

I tried manually setting the universe size by populating `max_workers` and used a simpler example.

```python
from mpi4py.futures import MPIPoolExecutor
from socket import gethostname

def compute(x):
    print(f"I am {gethostname()}!")
    return x * x

if __name__ == '__main__':
    with MPIPoolExecutor(max_workers=2, python_exe="python") as executor:
        results = executor.map(compute, range(10))
        for result in results:
            print(result)
```

But it seems the MPI backend is hitting an exception.

```
job aborted:
[ranks] message

[0] process exited without calling finalize

[1] terminated

---- error analysis -----

[0] on localhost
python ended prematurely and may have crashed. exit code 0xc0000409

---- error analysis -----
```

Any ideas how I can enable more verbose logging? I have already ran this with the most verbose logging Microsoft MPI can provide.

> Microsoft MPI have become abandonware. You should really try to switch to Intel MPI.

I would really love to but Intel MPI's Hydra Utility is closed-source and I am running on a AMD machine :/

**Reply by dalcinl** [reply]
> Any ideas how I can enable more verbose logging?

No, I'm not familiar enough with MS MPI as to suggest something. And as MS MPI seems no longer actively developed, I doubt you may be able to get help about it.

Are you able to execute a basic MPI helloworld in two different nodes? e.g. `mpiexec -n 2 -hosts ... python -m mpi4py.bench helloworld` ? If that works, then maybe you can try to execute mpi4py.futures in the legacy MPI-1 mode, read at the end of this subsection: https://mpi4py.readthedocs.io/en/stable/mpi4py.futures.html#computing-pi-parallel-task

> I would really love to but Intel MPI's Hydra Utility is closed-source and I am running on a AMD machine :/

Did you try? Is Intel actively preventing Intel MPI from working on AMD machines?

**Reply by winstxnhdw** [reply]
> Are you able to execute a basic MPI helloworld in two different nodes? e.g. mpiexec -n 2 -hosts ... python -m mpi4py.bench helloworld ? If that works, then maybe you can try to execute mpi4py.futures in the legacy MPI-1 mode, read at the end of this subsection: https://mpi4py.readthedocs.io/en/stable/mpi4py.futures.html#computing-pi-parallel-task

It stopped giving me the error but now it's only running on the other machine and returning a value of `0.0000` for the `cpi.py` test.

> Did you try? Is Intel actively preventing Intel MPI from working on AMD machines?

Not yet. I chose Microsoft MPI because I initially believed that to be the one that is most likely to work but given the amount of pain I am facing right now, I believe it is time to give Intel MPI a try.
